dir.rb

```diff
@@ -24,6 +24,97 @@ def self.open(name, encoding: nil, &block)
     end
   end
 
+  #
+  #  call-seq:
+  #     Dir.foreach( dirname ) {| filename | block }                 -> nil
+  #     Dir.foreach( dirname, encoding: enc ) {| filename | block }  -> nil
+  #     Dir.foreach( dirname )                                       -> an_enumerator
+  #     Dir.foreach( dirname, encoding: enc )                        -> an_enumerator
+  #
+  #  Calls the block once for each entry in the named directory, passing
+  #  the filename of each entry as a parameter to the block.
+  #
+  #  If no block is given, an enumerator is returned instead.
+  #
+  #     Dir.foreach("testdir") {|x| puts "Got #{x}" }
+  #
+  #  <em>produces:</em>
+  #
+  #     Got .
+  #     Got ..
+  #     Got config.h
+  #     Got main.rb
+  #
+  #
+  def self.foreach(path, encoding: Encoding.find("filesystem"))
+    return Enumerator.new(Dir.new(path, encoding: encoding)) if !block_given?
```

Comment:
`RETURN_ENUMERATOR` defers to open `path`.

### master
```shell
$ rm -fr xxx; ruby -e 'd = Dir.foreach("xxx"); Dir.mkdir("xxx"); p d.to_a'
[".", ".."]
```

### patched
```shell
$ rm -fr xxx; ./ruby -e 'd = Dir.foreach("xxx"); Dir.mkdir("xxx"); p d.to_a'
Traceback (most recent call last):
	3: from -e:1:in `<main>'
	2: from <internal:dir>:50:in `foreach'
	1: from <internal:dir>:50:in `new'
<internal:dir>:126:in `initialize': No such file or directory @ dir_initialize - xxx (Errno::ENOENT)
```

Reply:
Note that `Enumerator.new` yields a warning.

```shell
$ ruby -e 'Enumerator.new(Dir.new("."))'
-e:1: warning: Enumerator.new without a block is deprecated; use Object#to_enum instead
```

Reply:
Change code to using `RETURN_ENUMERATOR`.

Reply:
Passing `path` only makes `encoding:` option not working.
The enumerator should be returned in dir.rb, I guess.

Reply:
OK.
Fix this patch, like this.

```c
    VALUE argv[2];
    argv[0] = path;

    if (NIL_P(enc)) {
    RETURN_ENUMERATOR(io, 1, argv);
    } else {
    VALUE kw_hash = rb_hash_new();

    rb_gc_register_mark_object(kw_hash);
    rb_hash_aset(kw_hash, ID2SYM(rb_intern("encoding")), enc);
    rb_obj_freeze(kw_hash);
    argv[1] = kw_hash;

    RETURN_ENUMERATOR_KW(io, 2, argv, RB_PASS_KEYWORDS);
    }
```

Reply:
create `kw_hash`,  workaround for  `ArgumentError` 

```bash
$ rm -fr xxx; ruby -e 'd = Dir.foreach("xxx"); Dir.mkdir("xxx"); p d.to_a'
Traceback (most recent call last):
        2: from -e:1:in `<main>'
        1: from <internal:dir>:51:in `foreach'
<internal:dir>:15:in `open': No such file or directory @ dir_initialize - xxx (Errno::ENOENT)
```